File: benchmarks/reference/http_sim.js

class Request {
    constructor(method, path) {
        this.method = method;
        this.path = path;
    }
}

class Router {
    constructor() {
        self.routes = {};
    }

    add(method, path, handler) {
        let key = method + ":" + path;
        this.routes[key] = handler;
    }

    dispatch(req) {
        let key = req.method + ":" + req.path;
        if (this.routes[key]) {
            return this.routes[key]();
        }
        return "404 Not Found";
    }
}

function home() { return "Welcome Home"; }
function users() { return "User List"; }
function about() { return "About Us"; }

function main() {
    let router = new Router();
    router.routes = {}; // JS weirdness with 'self' in constructor copy-paste, fixing.
    
    // Fix: In JS class methods, 'this' is correct.
    // Re-defining properly below in actual write call.
}

// Correct JS Implementation
class RouterJS {
    constructor() {
        this.routes = {};
    }

    add(method, path, handler) {
        let key = method + ":" + path;
        this.routes[key] = handler;
    }

    dispatch(req) {
        let key = req.method + ":" + req.path;
        if (this.routes[key]) {
            return this.routes[key]();
        }
        return "404 Not Found";
    }
}

const start = process.hrtime();

const router = new RouterJS();
router.add("GET", "/", home);
router.add("GET", "/users", users);
router.add("GET", "/about", about);

for (let i = 0; i < 50000; i++) {
    let r1 = new Request("GET", "/");
    router.dispatch(r1);

    let r2 = new Request("GET", "/users");
    router.dispatch(r2);

    let r3 = new Request("GET", "/missing");
    router.dispatch(r3);
}

const diff = process.hrtime(start);
const elapsed = diff[0] + diff[1] / 1e9;
console.log(`HTTP Sim Time: ${elapsed.toFixed(6)}`);

File: benchmarks/reference/http_sim.go

package main

import (
	"fmt"
	"time"
)

type Request struct {
	Method string
	Path   string
}

type Router struct {
	Routes map[string]func() string
}

func NewRouter() *Router {
	return &Router{
		Routes: make(map[string]func() string),
	}
}

func (r *Router) Add(method, path string, handler func() string) {
	key := method + ":" + path
	r.Routes[key] = handler
}

func (r *Router) Dispatch(req Request) string {
	key := req.Method + ":" + req.Path
	if handler, ok := r.Routes[key]; ok {
		return handler()
	}
	return "404 Not Found"
}

func home() string { return "Welcome Home" }
func users() string { return "User List" }
func about() string { return "About Us" }

func main() {
	router := NewRouter()
	router.Add("GET", "/", home)
	router.Add("GET", "/users", users)
	router.Add("GET", "/about", about)

	start := time.Now()

	for i := 0; i < 50000; i++ {
		r1 := Request{Method: "GET", Path: "/"}
		router.Dispatch(r1)

		r2 := Request{Method: "GET", Path: "/users"}
		router.Dispatch(r2)

		r3 := Request{Method: "GET", Path: "/missing"}
		router.Dispatch(r3)
	}

	elapsed := time.Since(start).Seconds()
	fmt.Printf("HTTP Sim Time: %.6f\n", elapsed)
}
